src/unity/toolkits/tcmps/mps_device_manager.m

```diff
@@ -0,0 +1,96 @@
+//
+//  mps_device_manager.m
+//  tcmps
+//
+//  Copyright © 2018 Turi. All rights reserved.
+//
+
+#import "mps_device_manager.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface TCMPSDeviceManager ()
+{
+  // Token received from Metal to control notifications
+  id <NSObject> _deviceObserver;
+}
+
+// Internally re-declare this property as writable
+@property(readwrite, nullable, atomic) id <MTLDevice> preferredDevice;
+
+@end
+
+@implementation TCMPSDeviceManager
+
++ (instancetype)sharedInstance {
+  static TCMPSDeviceManager *singleton = nil;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    singleton = [[TCMPSDeviceManager alloc] init];
+  });
+  return singleton;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    // Define a notification handler that simply recomputes the preferred device
+    // from scratch on each change to the device list.
+    __weak TCMPSDeviceManager *weakSelf = self;
+    MTLDeviceNotificationHandler handler = ^(id <MTLDevice> device, MTLDeviceNotificationName name) {
+      [weakSelf setPreferredDeviceFromDevices:MTLCopyAllDevices()];
+      // (In theory we should only set self.preferredDevice on our own serial
+      // queue to guard against race conditions, but here it doesn't seem worth
+      // the cost in complexity.)
+    };
+
+    // Retrieve the initial device list and register for change notifications.
+    id <NSObject> deviceObserver = nil;
+    NSArray<id <MTLDevice>> *devices = MTLCopyAllDevicesWithObserver(&deviceObserver, handler);
+    _deviceObserver = deviceObserver;
+
+    // Set the initial preferred device.
+    [self setPreferredDeviceFromDevices:devices];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  MTLRemoveDeviceObserver(_deviceObserver);
+}
+
+- (BOOL)shouldPreferDevice:(id <MTLDevice>)candidate overDevice:(nullable id <MTLDevice>)current {
+  // Something is better than nothing.
+  if (!current) return YES;
```

Comment:
Not sure.  
Gustav's whole approach (based on some input from MPS, if I recall correctly) was that a low power GPU (which is usually just the onboard one) would perform slowly than the CPU.

Reply:
Well, FWIW, this code merely preserves the behavior defined in the original mps_dev.mm. Note that in practice MPS usage is gated on the Python side by `has_fast_mps_support`, which only returns true if there's a "high-power" Metal device.